src/haz3lcore/dynamics/Transition.re

```diff
@@ -0,0 +1,560 @@
+open Util;
+open PatternMatch;
+open DH;
+
+/* Transition.re
+
+   This module defines the evaluation semantics of Hazel in terms of small step
+   evaluation. These small steps are wrapped up into a big step in Evaluator.re.
+
+   I'll use the Sequence case as an example:
+
+    | Sequence(d1, d2) =>
+        let. _ = otherwise(d1 => Sequence(d1, d2))
+        and. _ = req_final(req(state, env), 0, d1);
+        Step({apply: () => d2, kind: Sequence, final: false});
+
+
+    Each step semantics starts with a `let. () = otherwise(...)` that defines how
+    to wrap the expression back up if the step couldn't be evaluated.
+
+    This is followed by a series of `and. d1' = req_final(req(state, env), <i>, <d1>)`
+    which indicate that in order to evaluate the step, <d1> must be final. (req_value
+    is also available if it needs to be a value). Note that if successful, d1' hold
+    contain the fully-evaluated version of d1.
+
+    Finally, we have the Step construct that defines the actual step. Note "Step"s
+    should be used if and only if they change the expression. If they do not change
+    the expression, use `Constructor` or `Indet`.
+
+    The step defines firstly, a `() => ...` function giving the result of the step,
+    secondly a `kind`, that describes the step (which will be used in the stepper)
```

Comment:
two reasons:
1. so that the test side effect happens at the right time
2. avoiding unexpected delays from "speculative execution" - perhaps not a problem atm (though there might be some big data structure cases where this is a problem); but when we start combining steps in the single-stepper this will be more of an issue